### PR TITLE
Throttle.Do does not panic after Finish

### LINKED ISSUE #1394: WriteBatch.Flush panics if called twice
I understand that flushing twice doesn't make sense, since a batch can only be flushed once, but the second call should error with a helpful message instead of panicking.

```
$ go version
go version devel +152ffca82f Mon Jun 29 08:06:32 2020 +0000 linux/amd64
$ go list -m github.com/dgraph-io/badger/v2
github.com/dgraph-io/badger/v2 v2.0.3
$ cat main.go
package main

import (
	"io/ioutil"
	"os"

	"github.com/dgraph-io/badger/v2"
)

func main() {
	dir, err := ioutil.TempDir("", "badger-test")
	if err != nil {
		panic(err)
	}
	defer os.RemoveAll(dir)
	db, err := badger.Open(badger.DefaultOptions(dir))
	if err != nil {
		panic(err)
	}
	defer db.Close()

	wb := db.NewWriteBatch()
	wb.Flush()
	wb.Flush()
}
$ go run main.go
badger 2020/06/30 18:36:15 INFO: All 0 tables opened in 0s
badger 2020/06/30 18:36:15 INFO: Got compaction priority: {level:0 score:1.73 dropPrefix:[]}
panic: send on closed channel

goroutine 1 [running]:
github.com/dgraph-io/badger/v2/y.(*Throttle).Do(0xc0000200c0, 0x6e7260, 0xc0000200c0)
	/home/mvdan/go/pkg/mod/github.com/dgraph-io/badger/v2@v2.0.3/y/y.go:260 +0xc9
github.com/dgraph-io/badger/v2.(*WriteBatch).commit(0xc000020180, 0x0, 0x0)
	/home/mvdan/go/pkg/mod/github.com/dgraph-io/badger/v2@v2.0.3/batch.go:141 +0x45
github.com/dgraph-io/badger/v2.(*WriteBatch).Flush(0xc000020180, 0x0, 0x0)
	/home/mvdan/go/pkg/mod/github.com/dgraph-io/badger/v2@v2.0.3/batch.go:155 +0x45
main.main()
	/home/mvdan/src/test/main.go:24 +0x1fa
exit status 2
```

### PR DESCRIPTION
**Pull request by ekalinin**
Fixes #1394

I'm not sure about an error at `Do`. Maybe it will be ok to just return `nil`.

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/badger/1396)

### CODE REVIEW

**Comment by ekalinin** [comment]
Thanks for comments @jarifibrahim 
Fixed all your comments. Please, take a look.

**Comment by ekalinin** [comment]
> Some minor comments. Please address them and this PR is good to go 
> 
> 
> It should be enough to do a blind update. My previous comment was incorrect. Apologies.
> So we can remove the `ErrDoubleFinish` too.
> 

Done.

> 
> You can remove the require.NotPanics.
> 
> A simple require.Error should be enough.

Done.

File: y/y.go [hunk 1]
```diff
@@ -267,6 +271,9 @@ func NewThrottle(max int) *Throttle {
 // are already maximum number of workers working. If it detects an error from
 // previously Done workers, it would return it.
 func (t *Throttle) Do() error {
+	if t.finished {
```
Comment (NamanJain8) [inline y/y.go:274]: @ekalinin, consider a scenario where Do() and Finish() are called by different goroutines. In that case, there is a race condition. 
Thread 1 passes this check while thread 2 closes the channel and then thread 1 push into closed channel `ch`.
Try running this scenario using --race flag.

**Comment by NamanJain8** [comment]
@ekalinin, In the Finish(), you can have an atomic Store operation for `finished`. While in Do(), you can use a lock to ensure that Finish() does not update the `finished` while Do() is updating executing.

**Comment by ekalinin** [comment]
> @ekalinin, In the Finish(), you can have an atomic Store operation for `finished`. While in Do(), you can use a lock to ensure that Finish() does not update the `finished` while Do() is updating executing.

Thanks for the hint, @NamanJain8!
Fixed.

File: y/y.go [hunk 2]
```diff
@@ -41,6 +41,9 @@ var (
 	// compression algorithm is being used for compression. ZSTD cannot work
 	// without CGO.
 	ErrZstdCgo = errors.New("zstd compression requires building badger with cgo enabled")
+
+	// ErrDoAfterFinish indicates a pointless Do call after Finish
+	ErrDoAfterFinish = errors.New("Do has no any point after Finish")
```
Comment (NamanJain8) [inline y/y.go:46]: Looks good to me. 🥳 
Can you just rephrase this to something like "Do should not be called after Finish"?
@jarifibrahim what do you say?

Comment (ekalinin) [reply y/y.go]: Done.

**Comment by jarifibrahim** [comment]
Hey @ekalinin, we got this PR reviewed internally and there are a few changes needed. @NamanJain8 will make those change and push on this PR. Thank you for helping fix this :tada:

**Comment by ekalinin** [comment]
> Hey @ekalinin, we got this PR reviewed internally and there are a few changes needed. @NamanJain8 will make those change and push on this PR. Thank you for helping fix this

Cool! Thanks for help as well :) !

File: batch_test.go [hunk 3]
```diff
@@ -123,3 +125,22 @@ func TestEmptyWriteBatch(t *testing.T) {
...
+			wb := db.NewWriteBatch()
+			wb.Flush()
+			require.Error(t, y.ErrCommitAfterFinish, func() { wb.Flush() })
+		})
+	})
+	t.Run("flush after cancel", func(t *testing.T) {
+		runBadgerTest(t, nil, func(t *testing.T, db *DB) {
+			wb := db.NewWriteBatch()
+			wb.Cancel()
+			require.Error(t, y.ErrCommitAfterFinish, func() { wb.Flush() })
```
Comment (NamanJain8) [inline batch_test.go:143]: Remove func

File: batch.go [hunk 4]
```diff
@@ -194,8 +201,11 @@ func (wb *WriteBatch) commit() error {
 // returns any error stored by WriteBatch.
 func (wb *WriteBatch) Flush() error {
 	wb.Lock()
-	// commit will set the wb.err so no need to check the error here.
-	_ = wb.commit()
+	err := wb.commit()
+	if err != nil {
+		return err
```
Comment (NamanJain8) [inline batch.go:206]: Unlock()
